### ISSUE TITLE
Scraper seems to hang on download

### ISSUE THREAD

**Issue by opaquelytransparent** [DIGITALCRIMINALS/UltimaScraper#902]
Using the scraper, I'm able to get beyond selecting profile as well as which user to scrape, but scraper appears as below and no progress seems to be made. I've tried leaving the script running for upwards of 30 minutes without any appearance of movement or downloads. Here is the text of what the script does:

PS C:\OnlyFans - DIGITALCRIMINALS> poetry run python start_us.py
Choose Sites: 0 = All | 1 = OnlyFans | 2 = Fansly
1
Choose Profiles: 0 = All | 1 = default
1
Choose Subscriptions: 0 = All | 1 = backandbetter | 2 = utahjaz | 3 = kenzieaaliyahxo | 4 = laurenmochen | 5 = nataliereynoldss | 6 = doctorpov | 7 = thehannistons | 8 = justjenniferh | 9 = butternutgiraffe | 10 = keelingz | 11 = epdtravels | 12 = gracecharisxo | 13 = jadeteen | 14 = savvysuxx | 15 = alexandramhunt | 16 = mikaelapascal
0
Processing Scraped Posts
100%|██████████████████████████████████████████████████████████████████████████████| 187/187 [00:00<00:00, 2372.90it/s]
Processing metadata.
Finished processing metadata.
Renaming files.
100%|██████████████████████████████████████████████████████████████████████████████| 188/188 [00:00<00:00, 2436.88it/s]
Processing Scraped Messages
100%|███████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 471.75it/s]
Processing metadata.
Finished processing metadata.
Renaming files.
100%|██████████████████████████████████████████████████████████████████████████████████| 6/6 [00:00<00:00, 2054.35it/s]
Download Processing
Name: laurenmochen | Type: Posts | Downloading: 3 | Total: 3 | Duplicates: 0  | Directory: __user_data__\sites

**Comment by wooki0** [comment]
are you ip banned try logging on in a browser i can't get this far it onlyfans temp bans my ip

**Comment by divakar123** [comment]
same issue

**Comment by wooki0** [comment]
> same issue

Yeah every time i try to start the ripper its pretty much an instant ban for me won't go past me selecting a profile option

**Comment by CannotTouch** [comment]
We are talking about it here: https://github.com/DIGITALCRIMINALS/OnlyFans/issues/838
The only suggest for now is low the max thread, personally i'm at 2

**Comment by sweetcatastrophe** [comment]
> We are talking about it here: #838 The only suggest for now is low the max thread, personally i'm at 2

I'm having the same issue and want to test the max thread thing, but where is it? I'm using the latest version but I don't have it in my config file

**Comment by CannotTouch** [comment]
line 8 of the version 8.0 of the config file.

**Comment by sweetcatastrophe** [comment]
> line 8 of the version 8.0 of the config file.

Oh I'm dumb, I was a looking at a different part of the config. Got it now, thank you.

Still having the issue though unfortunately, gets stuck at "Download Processing", or it just does nothing after I select a subscription to download.

**Comment by avekifes** [comment]
I had the same issue, but I wasn't banned...or maybe I was, but didn't realize it because I didn't check the site until later...

**Comment by wooki0** [comment]
> I had the same issue, but I wasn't banned...or maybe I was, but didn't realize it because I didn't check the site until later...

yeah its a soft ban it lets up after a few mins most of the time but makes the ripper unusable if you have a boat load of subs

**Comment by opaquelytransparent** [comment]
Even with changing max threads to 1 or 2, still hangs on the "download processing" as in my original post. I am still able to log into OnlyFans regardless. Even ripping only 1 sub at a time causes it to hang

**Comment by ChillyP** [comment]
Having the same issue, scraper seems to hang and it’s not due to a 429 error.

**Comment by dvick** [comment]
Same here. Scraper keeps hanging randomly, but I'm not getting 429'd—unless it's lasting a very short time now and the scraper's just not recovering. Was working no problem with max threads at 3 before updating a day or two ago.

**Comment by webhead05** [comment]
Same issue for me as well. I've turned messages to "false" otherwise I will get the soft ban. After that though, if I choose "0=All" for my chosen model, it hangs on "Messages". If I choose "1=Images", it looks like it downloads Messages and then hangs on Imgaes. It did, however, download the newest image. It's not downloading all images when there are multiple in a message or post though.

**Comment by EvilDragonfiend** [comment]
I have the same problem. which python version are you using? I haven't tried it on `3.10.1`, but mine is `3.11`. maybe it'd be because of it.

**Comment by ChillyP** [comment]
I'm using 3.10.9, tried just getting posts and it still hangs after renaming files.

**Comment by Bearded-Beard** [comment]
So I had to laugh, ironically I'm subbed to this creator and mine hangs after the first one, it was working perfectly before I updated to the new system now I can't download anything, now, I'm still able to access OF while I'm trying to download, and I've let it stay on 12hrs to see if its going to dl, I've also tried the turn the messages to false aswell and same thing, does one profile and hangs.

**Comment by dairyroach** [comment]
Having the same problem.  I don't get the 429 error and I have my threads set to 5.  Still hangs.  Here's the traceback I get when I interrupt the scraper:

`Download Processing
Name: jesse_switch | Type: Posts | Downloading: 57 | Total: 57 | Duplicates: 0  | Directory: \\Beedle\Backpack\backup

Traceback (most recent call last):
  File "C:\Users\richa\Downloads\OnlyFans-master-new\scraper\start_us.py", line 71, in <module>
    asyncio.run(main())
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1871, in _run_once
    event_list = self._selector.select(timeout)
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 444, in select
    self._poll(timeout)
  File "C:\Users\richa\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 797, in _poll
    status = _overlapped.GetQueuedCompletionStatus(self._iocp, ms)`

**Comment by dairyroach** [comment]
Having the same problem.  I don't get the 429 error and can still log on just fine in browser.  Here's the traceback I get when I interrupt the scraper:

`Download Processing
Name: jesse_switch | Type: Posts | Downloading: 57 | Total: 57 | Duplicates: 0  | Directory: \\Beedle\Backpack\backup

Traceback (most recent call last):
  File "C:\Users\user\Downloads\OnlyFans-master-new\scraper\start_us.py", line 71, in <module>
    asyncio.run(main())
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1871, in _run_once
    event_list = self._selector.select(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 444, in select
    self._poll(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 797, in _poll
    status = _overlapped.GetQueuedCompletionStatus(self._iocp, ms)`

**Comment by dairyroach** [comment]
Having the same problem.  I don't get the 429 error and can still log on just fine in browser.  It only seems to happen on this one model.  Here's the traceback I get when I interrupt the scraper:

```
Download Processing
Name: jesse_switch | Type: Posts | Downloading: 57 | Total: 57 | Duplicates: 0  | Directory: C:\Users\user\Downloads\OnlyFans-master-new\download

Traceback (most recent call last):
  File "C:\Users\user\Downloads\OnlyFans-master-new\scraper\start_us.py", line 71, in <module>
    asyncio.run(main())
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1871, in _run_once
    event_list = self._selector.select(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 444, in select
    self._poll(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 797, in _poll
    status = _overlapped.GetQueuedCompletionStatus(self._iocp, ms)
```

**Comment by ChillyP** [comment]
Mine never makes it to download processing, here's the traceback.

Finished processing metadata.
Renaming files.
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6113/6113 [4:05<00:00, 24.95it/s]
Traceback (most recent call last):
  File "C:\Users\user\Documents\GitHub\OnlyFans\start_us.py", line 71, in <module>
    asyncio.run(main())
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 603, in run_forever
    self._run_once()
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1868, in _run_once
    event_list = self._selector.select(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 439, in select
    self._poll(timeout)
  File "C:\Users\user\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 792, in _poll
    status = _overlapped.GetQueuedCompletionStatus(self._iocp, ms)

**Comment by Dragoviano** [comment]
Same here. It only downloads one single file, be it img or video, then it hangs. Then I have to run the script again to download a new single file. I have already tested with 1, 3, 5, 10 and 32 max threads, all the same.

**Comment by twist3dimages** [comment]
> Same here. It only downloads one single file, be it img or video, then it hangs. Then I have to run the script again to download a new single file. I have already tested with 1, 3, 5, 10 and 32 max threads, all the same.

how do you modify the number of threads in the new version

**Comment by Dragoviano** [comment]
> > Same here. It only downloads one single file, be it img or video, then it hangs. Then I have to run the script again to download a new single file. I have already tested with 1, 3, 5, 10 and 32 max threads, all the same.
> 
> how do you modify the number of threads in the new version

Just go to __settings__ folder, open config file with a notepad, it's on line 8.

**Comment by god43** [comment]
same issue, mine seems to just hang as well, not sure why though?

**Comment by Pwnager12345** [comment]
Mine is going the exact same thing but only on certain profiles. The rest are downloading fine.

**Comment by Bearded-Beard** [comment]
I've had to go back to a version that doesn't have ultimate scraper and it works for now.

**Comment by ChillyP** [comment]
> Mine is going the exact same thing but only on certain profiles. The rest are downloading fine.

Which build did you go back to?

**Comment by twist3dimages** [comment]
> Mine is going the exact same thing but only on certain profiles. The rest are downloading fine.

same

**Comment by loli-ryona** [comment]
Current work around I have for this issue is using the 7.6.1 release.
Just used existing `auth.json` from the lastest version.
It will crash when scraping stories so you need to open `modules/onlyfans.py` and comment out line **781** to **790**.
After doing the it worked fine.

**Comment by Bearded-Beard** [comment]
> Current work around I have for this issue is using the 7.6.1 release. Just used existing `auth.json` from the lastest version. It will crash when scraping stories so you need to open `modules/onlyfans.py` and comment out line **781** to **790**. After doing the it worked fine.

I've tried that, but I cop a "Can't locate revision identified by 'set of numbers and letters'" and everything I google doesn't seem to help

**Comment by excludedBittern8** [comment]
Since I was having so much trouble with this script. I decided to dive into making my own version.
I'm only focusing on onlyfans which makes things a little easier to manage.
I have 400+ historical subscriptions. Many of which still send ppv through the inbox.
Maybe 100 free subscriptions.

https://github.com/excludedBittern8/ofscraper

I've added features for better file organization. Something more similar to this script
We will also be added a feature to increase repeated scraping speed 
basically if you scrape a model once, the second time will be much fsster

**Comment by ChillyP** [comment]
> EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working

What did you have to do to the DB to get it working??

**Comment by Bearded-Beard** [comment]
> > EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working
> 
> What did you have to do to the DB to get it working??

https://github.com/DIGITALCRIMINALS/UltimaScraperCollection/tree/04ec980f0fb1609ae6b8ddf61c59443b41ac4004/ultima_scraper_collection/managers/database_manager/connections/sqlite/databases/user_data/migration/alembic/versions copy these into the db directory in the nov16 version and it should work

**Comment by betoalanis** [comment]
> EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working

wym Nov16 commit? can't find a commit on that day. Also, this means undoing the Jan23 migration? so a new install with the old commit you say, but bump those specific files to the newest version?

**Comment by Bearded-Beard** [comment]
https://github.com/DIGITALCRIMINALS/OnlyFans/tree/aa8bc84e39922c7c33ca970d2d731bf9696092af

**Comment by betoalanis** [comment]
> https://github.com/DIGITALCRIMINALS/OnlyFans/tree/aa8bc84e39922c7c33ca970d2d731bf9696092af

thank you :) time zones... shows nov15 for me, just wanted to be sure.

**Comment by dairyroach** [comment]
> > > EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working
> > 
> > 
> > What did you have to do to the DB to get it working??
> 
> https://github.com/DIGITALCRIMINALS/UltimaScraperCollection/tree/04ec980f0fb1609ae6b8ddf61c59443b41ac4004/ultima_scraper_collection/managers/database_manager/connections/sqlite/databases/user_data/migration/alembic/versions copy these into the db directory in the nov16 version and it should work

Which directory exactly are you putting those files in?  They seemed to belong in here, but wasn't sure if this was what you were talking about:

\database\databases\user_data\alembic\versions

Was there any other step besides copying the files into the db folder?  I keep getting the error, "invalid decimal literal (0d4d92c0498e_content.py, line 75)"

**Comment by god43** [comment]
> > > EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working
> > 
> > 
> > What did you have to do to the DB to get it working??
> 
> https://github.com/DIGITALCRIMINALS/UltimaScraperCollection/tree/04ec980f0fb1609ae6b8ddf61c59443b41ac4004/ultima_scraper_collection/managers/database_manager/connections/sqlite/databases/user_data/migration/alembic/versions copy these into the db directory in the nov16 version and it should work

that commit works anyways, i didn't have to copy db files in the first place? why would we need to do this?

**Comment by jdkjdji** [comment]
Hello, I use mac, so... i don't know if the solution works in windows. but, i'm install in terminal postgresql, and run again  python updater.py  and works...

**Comment by Bearded-Beard** [comment]
> > > > EDIT: Nov16 commit with some .db alemic (sp) files, is working for me for now, slow but working
> > > 
> > > 
> > > What did you have to do to the DB to get it working??
> > 
> > 
> > https://github.com/DIGITALCRIMINALS/UltimaScraperCollection/tree/04ec980f0fb1609ae6b8ddf61c59443b41ac4004/ultima_scraper_collection/managers/database_manager/connections/sqlite/databases/user_data/migration/alembic/versions copy these into the db directory in the nov16 version and it should work
> 
> that commit works anyways, i didn't have to copy db files in the first place? why would we need to do this?

It doesn't work for everyone thats why I looked for the answer for the issue I had, and was offering others help if they got that issue aswell...

**Comment by N0m3nN3sc10** [comment]
I ran the scraper through a debugger and found that (on Windows at least) this hanging issue seems to be down to the `format_image()` code starting at line 111 of main_helper.py.

It's going into an infinite loop after trying and failing to import `win32_setctime` at line 116, thus generating a `ModuleNotFoundError("No module named 'win32_setctime'")` exception. However, the exception handler at line 123 just jumps back to the beginning to keep retrying the import endlessly.

Running a `poetry add win32-setctime` command from the project folder installs the missing dependency which allows the code to break out of the loop and continue downloading.

**Comment by dairyroach** [comment]
> I ran the scraper through a debugger and found that (on Windows at least) this hanging issue seems to be down to the `format_image()` code starting at line 111 of main_helper.py.
> 
> It's going into an infinite loop after trying and failing to import `win32_setctime` at line 116, thus generating a `ModuleNotFoundError("No module named 'win32_setctime'")` exception. However, the exception handler at line 123 just jumps back to the beginning to keep retrying the import endlessly.
> 
> Running a `poetry add win32-setctime` command from the project folder installs the missing dependency which allows the code to break out of the loop and continue downloading.

This fixed the problem for me!  At least, in the November 16th build.  The latest UltimaScraper version still just hangs.

Thank you, kind hero!  I will forever tell the tale of what you did for us here today.

**Comment by N0m3nN3sc10** [comment]
> This fixed the problem for me! At least, in the November 16th build. The latest UltimaScraper version still just hangs.

I'm running the latest commit just fine now, but I had previously deleted the original venv created by poetry and reinstalled all of the dependencies in a new venv, so you might like to give that a try.

By default it'll be somewhere in `%LocalAppData%\pypoetry\Cache\virtualenvs`, so just find it there, delete it, and run `poetry install` from the project folder to recreate it.

You can also make it easier to reset the venv in future by configuring poetry to always create the venv inside the project folder with `poetry config virtualenvs.in-project true` before you run `poetry install`. That way you keep the project and its dependencies all in one place.

**Comment by Arko109** [comment]
For me, the `poetry add win32-setctime` on the latest master commit solved the issue, the script downloads again.

**Comment by betoalanis** [comment]
I made it work and was able to DL about 50GB of content  by reinstalling everything after a problem with my computer and needing to reformat my drive. And by everything I mean EVERYTHING, i started from python.. and step by step installing each thing that the script was telling me was missing, **step by step**, meaning didn't install anything I thought I needed until the script itself installed it or asked for it.

also, i lowered my max thread to 2.

I don't mean for you guys to reformat your drives, but if possible, uninstall all the dependencies and start from scratch.
